src/java.base/share/classes/java/lang/Thread.java

```diff
@@ -782,29 +772,22 @@ public Thread(ThreadGroup group, Runnable target, String name,
      * @see        #stop()
      */
     public synchronized void start() {
-        /**
-         * This method is not invoked for the main method thread or "system"
-         * group threads created/set up by the VM. Any new functionality added
-         * to this method in the future may have to also be added to the VM.
-         *
-         * A zero status value corresponds to state "NEW".
-         */
-        if (threadStatus != 0)
-            throw new IllegalThreadStateException();
-
-        /* Notify the group that this thread is about to be started
-         * so that it can be added to the group's list of threads
-         * and the group's unstarted count can be decremented. */
-        group.add(this);
+        boolean success = false;
 
-        boolean started = false;
         try {
-            start0();
-            started = true;
+            synchronized (blockerLock) {
+                if (started) {
+                    throw new IllegalThreadStateException();
+                }
+
+                group.add(this);
+                startImpl();
+                success = true;
+            }
         } finally {
             try {
-                if (!started) {
-                    group.threadStartFailed(this);
+                if (!success && !started) {
+                    group.threadTerminated(this);
```

Comment:
I think this should call `threadStartFailed()` instead.

Reply:
`threadStartFailed ` only removes the thread from the group and update `nUnstartedThreads`, which seems not enough for `OpenJ9` failed thread.
iirc there was a hang in a test,  using `threadTerminated()` invokes `notifyAll()` when `nthreads == 0` which solved the issue. There is another similar `OpenJ9` change [1] as well.

[1] https://github.com/eclipse-openj9/openj9/blob/a8242cf13af02a7cdb11218c805163fa9f4140d8/jcl/src/java.base/share/classes/java/lang/J9VMInternals.java#L323

Reply:
This is different: we're talking about a thread that failed to start.

Reply:
The OpenJ9 change was introduced at same time as this PR to solve the test failure.
Currently I am not able to reproduce the test hang w/o it, there are lots of known segmentation error in JDK19 builds such as https://github.com/eclipse-openj9/openj9/issues/14228.

Reply:
I don't know which failure modes you're referring to. None of what I see in eclipse-openj9/openj9#14228 seems related to `Thread.start()`.

Reply:
This is a different one, not existing OpenJ9 behaviours.
Initial RI code uses `threadStartFailed()` which caused a test hang, `threadTerminated` fixed the failure (iirc one thread was waiting for this failed thread, `notifyAll()` awoke it, this probably was due to OpenJ9 specific thread lifetime). However I can't reproduce the initial hang w/o this change as per https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/387#discussion_r788809418.

Do you see any problem w/ usage of `threadTerminated()`?

Reply:
`threadStartFailed()` is the right method to call if `start()` fails, but a constructed, but unstarted thread must also call `ThreadGroup.addUnstarted()` to balance the bookkeeping.

Reply:
Use `threadStartFailed()` and verified w/ personal builds.